File: dawn_python/ss.py
from collections import Counter

def get_ngrams(n, s):
    return [s[i:i+n] for i in range(len(s) - n + 1)]

def count_3gram_frequencies(sample_keys):
    frequencies = Counter()
    for key in sample_keys:
        frequencies.update(get_ngrams(3, key))
    return frequencies

def select_3gram_intervals(sample_keys, dict_size):
    frequencies = count_3gram_frequencies(sample_keys)
    sorted_frequencies = sorted(frequencies.items(), key=lambda x: x[1], reverse=True)
    selected_ngrams = [ngram for ngram, _ in sorted_frequencies[:dict_size // 2]]
    selected_ngrams.sort()

    intervals = []
    last_ngram = ""
    for ngram in selected_ngrams:
        start = last_ngram
        if start:
            start = start[:-1] + chr(ord(start[-1]) + 1)
        intervals.append((start, ngram))
        last_ngram = ngram

    return intervals

File: dawn_haskell/src/Ss.hs
module SymbolSelector where

import qualified Data.List as L
import qualified Data.Map as M

get_ngrams n s = [take n (drop i s) | i <- [0..(length s - n + 1)]]

count_3gram_frequencies sample_keys = M.fromListWith (+) [(ngram, 1) | key <- sample_keys, ngram <- get_ngrams 3 key]

select_3gram_intervals sample_keys dict_size =
  let frequencies = count_3gram_frequencies sample_keys
      sorted_frequencies = L.sortBy (\(_, freq1) (_, freq2) -> compare freq2 freq1) $ M.toList frequencies
      selected_ngrams = map fst $ take (dict_size `div` 2) sorted_frequencies
      intervals = foldr (\ngram acc -> case acc of
                           [] -> [(ngram, Nothing)]
                           (last_ngram, _):rest -> let start = last_ngram ++ [chr $ ord (last last_ngram) + 1]
                                                    in (start, Just ngram):acc) [] $ L.sort selected_ngrams
  in intervals